browser/ntp_sponsored_images/new_tab_page_branded_view_counter.cc

```diff
@@ -0,0 +1,206 @@
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+#include "components/keyed_service/content/browser_context_keyed_service_factory.h"
+#include "components/prefs/pref_change_registrar.h"
+#include "components/prefs/pref_service.h"
+
+namespace {
+
+std::unique_ptr<NTPSponsoredImagesData> GetDemoWallpaper() {
+  auto demo = std::make_unique<NTPSponsoredImagesData>();
+  demo->wallpaper_image_urls = {
+      "ntp-dummy-brandedwallpaper-background-1.jpg",
+      "ntp-dummy-brandedwallpaper-background-2.jpg",
+      "ntp-dummy-brandedwallpaper-background-3.jpg"};
+  demo->logo_image_url = "ntp-dummy-brandedwallpaper-logo.png";
+  demo->logo_alt_text = "Technikke: For music lovers.";
+  demo->logo_company_name = "Technikke";
+  demo->logo_destination_url = "https://brave.com";
+  return demo;
+}
+
+class NewTabPageBrandedViewCounterFactory
+    : public BrowserContextKeyedServiceFactory {
+ public:
+  static NewTabPageBrandedViewCounter* GetForProfile(Profile* profile) {
+    return static_cast<NewTabPageBrandedViewCounter*>(
+        GetInstance()->GetServiceForBrowserContext(profile, true));
+  }
+
+  static NewTabPageBrandedViewCounterFactory* GetInstance() {
+    static base::NoDestructor<NewTabPageBrandedViewCounterFactory> instance;
+    return instance.get();
+  }
+  NewTabPageBrandedViewCounterFactory() : BrowserContextKeyedServiceFactory(
+        "NewTabPageBrandedViewCounter",
+        BrowserContextDependencyManager::GetInstance()) {}
+  ~NewTabPageBrandedViewCounterFactory() override = default;
+
+  NewTabPageBrandedViewCounterFactory(
+      const NewTabPageBrandedViewCounterFactory&) = delete;
+  NewTabPageBrandedViewCounterFactory& operator=(
+      const NewTabPageBrandedViewCounterFactory&) = delete;
+
+ private:
+  // BrowserContextKeyedServiceFactory:
+  KeyedService* BuildServiceInstanceFor(
+      content::BrowserContext* browser_context) const override {
+    NewTabPageBrandedViewCounter* instance = new NewTabPageBrandedViewCounter(
+        Profile::FromBrowserContext(browser_context));
+    return instance;
+  }
+  content::BrowserContext* GetBrowserContextToUse(
+      content::BrowserContext* context) const override {
+    return chrome::GetBrowserContextRedirectedInIncognito(context);
+  }
+};
+
+}  // namespace
+
+// static
+NewTabPageBrandedViewCounter* NewTabPageBrandedViewCounter::GetForProfile(
+    Profile* profile) {
+  if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaper))
+    return NewTabPageBrandedViewCounterFactory::GetForProfile(profile);
+  return nullptr;
+}
+
+NewTabPageBrandedViewCounter::NewTabPageBrandedViewCounter(Profile* profile)
+    : profile_(profile) {
+  // If we have a wallpaper, store it as private var.
+  // Set demo wallpaper if a flag is set.
+  if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo)) {
+    current_wallpaper_ = GetDemoWallpaper();
+  } else {
+    NTPSponsoredImagesComponentManager* manager =
```

Comment:
this should be connected to the NTPSponsoredImageService I think with its own observer for when sponsored images are available